### PR TITLE
daemon: limit number of concurrent healthchecks

### PR DESCRIPTION
**Pull request by stevvooe**
Signed-off-by: Stephen J Day <stephen.day@docker.com>

Attempt to mitigate problems with healthchecks seen in https://github.com/moby/moby/issues/36661, https://github.com/moby/moby/issues/33933 and others.

### CODE REVIEW

**Review by mptmpt096 — APPROVED** [review]

**Comment by stevvooe** [comment]
Confirmed that this is addressing the issues in https://github.com/moby/moby/issues/36661 with https://github.com/moby/moby/issues/36661#issuecomment-378045186. Cleaning up this PR to remove debugging messages.

**Comment by stevvooe** [comment]
PR has been cleaned up and now we have a test.

File: daemon/daemon.go [hunk 1]
```diff
@@ -858,6 +860,7 @@ func NewDaemon(config *config.Config, registryService registry.Service, containe
 	d.execCommands = exec.NewStore()
 	d.idIndex = truncindex.NewTruncIndex([]string{})
 	d.statsCollector = d.newStatsCollector(1 * time.Second)
+	d.healthcheckLimiter = newLimiter(2 * runtime.NumCPU()) // simple hueristic for concurrent execs
```
Comment (fcrisciani) [inline daemon/daemon.go]: hueristic --> heuristic

File: daemon/limiter.go [hunk 2]
```diff
@@ -0,0 +1,78 @@
+package daemon
+
+import (
+	"sync"
+)
+
+// limiter limits the number of concurrent goroutines running at a time by
```
Comment (fcrisciani) [inline daemon/limiter.go]: why not using a weighted semaphore: https://github.com/golang/sync/blob/master/semaphore/semaphore.go ?

**Comment by crosbymichael** [comment]
LGTM

File: daemon/limiter.go (hunk 2 above)
Comment (stevvooe) [reply daemon/limiter.go]: This implementation is indeed a weighted semaphore using channel coordination. I was not aware they had added a proper semaphore implementation.

I'll update this PR to use the semaphore version, as that seems more efficient.

**Comment by jahkeup** [comment]
Looks like `golang.org/x/sync/semaphore` is a new dependency and failing to build.

**Review by thaJeztah — APPROVED** [review]
LGTM

File: daemon/health.go [hunk 3]
```diff
@@ -188,12 +188,19 @@ func handleProbeResult(d *Daemon, c *container.Container, result *types.Healthch
...
 	probeTimeout := timeoutWithDefault(c.Config.Healthcheck.Timeout, defaultProbeTimeout)
 	probeInterval := timeoutWithDefault(c.Config.Healthcheck.Interval, defaultProbeInterval)
+loop:
 	for {
 		select {
 		case <-stop:
 			logrus.Debugf("Stop healthcheck monitoring for container %s (received while idle)", c.ID)
 			return
 		case <-time.After(probeInterval):
+			if err := d.healthcheckLimiter.Acquire(context.Background(), 1); err != nil {
```
Comment (fcrisciani) [inline daemon/health.go]: context has to be cancellable else the Acquire will remain stuck forever

Comment (stevvooe) [reply daemon/health.go]: Which context do we use for that?

Comment (stevvooe) [reply daemon/health.go]: Do we want to move the timeout up, above the acquire?

Comment (stevvooe) [reply daemon/health.go]: Moved the timeout up to the top. I'm going to add metrics to track the health check runtime and waiting period.

**Comment by stevvooe** [comment]
I've add another commit with some new health check counters that can help us to understand the behavior under load. The output is the following:

```
# TYPE engine_daemon_health_checks_active gauge
engine_daemon_health_checks_active 16
# HELP engine_daemon_health_checks_duration_seconds Time spent waiting, executing, and registering the result of a single healthcheck
# TYPE engine_daemon_health_checks_duration_seconds histogram
engine_daemon_health_checks_duration_seconds_bucket{le="0.005"} 0
engine_daemon_health_checks_duration_seconds_bucket{le="0.01"} 0
engine_daemon_health_checks_duration_seconds_bucket{le="0.025"} 0
engine_daemon_health_checks_duration_seconds_bucket{le="0.05"} 0
engine_daemon_health_checks_duration_seconds_bucket{le="0.1"} 0
engine_daemon_health_checks_duration_seconds_bucket{le="0.25"} 0
engine_daemon_health_checks_duration_seconds_bucket{le="0.5"} 0
engine_daemon_health_checks_duration_seconds_bucket{le="1"} 0
engine_daemon_health_checks_duration_seconds_bucket{le="2.5"} 374
engine_daemon_health_checks_duration_seconds_bucket{le="5"} 424
engine_daemon_health_checks_duration_seconds_bucket{le="10"} 424
engine_daemon_health_checks_duration_seconds_bucket{le="+Inf"} 424
engine_daemon_health_checks_duration_seconds_sum 737.1056642489999
engine_daemon_health_checks_duration_seconds_count 424
# HELP engine_daemon_health_checks_failed_total The total number of failed health checks
# TYPE engine_daemon_health_checks_failed_total counter
engine_daemon_health_checks_failed_total 0
# HELP engine_daemon_health_checks_total The total number of health checks
# TYPE engine_daemon_health_checks_total counter
engine_daemon_health_checks_total 440
# HELP engine_daemon_health_checks_wait_duration_seconds Time spent waiting to acquire health check semaphore
# TYPE engine_daemon_health_checks_wait_duration_seconds histogram
engine_daemon_health_checks_wait_duration_seconds_bucket{le="0.005"} 121
engine_daemon_health_checks_wait_duration_seconds_bucket{le="0.01"} 122
engine_daemon_health_checks_wait_duration_seconds_bucket{le="0.025"} 126
engine_daemon_health_checks_wait_duration_seconds_bucket{le="0.05"} 136
engine_daemon_health_checks_wait_duration_seconds_bucket{le="0.1"} 149
engine_daemon_health_checks_wait_duration_seconds_bucket{le="0.25"} 174
engine_daemon_health_checks_wait_duration_seconds_bucket{le="0.5"} 202
engine_daemon_health_checks_wait_duration_seconds_bucket{le="1"} 259
engine_daemon_health_checks_wait_duration_seconds_bucket{le="2.5"} 440
engine_daemon_health_checks_wait_duration_seconds_bucket{le="5"} 440
engine_daemon_health_checks_wait_duration_seconds_bucket{le="10"} 440
engine_daemon_health_checks_wait_duration_seconds_bucket{le="+Inf"} 440
engine_daemon_health_checks_wait_duration_seconds_sum 297.814396283
engine_daemon_health_checks_wait_duration_seconds_count 440
# HELP engine_daemon_health_checks_waiters The number of health checks waiting to run
# TYPE engine_daemon_health_checks_waiters gauge
engine_daemon_health_checks_waiters 22
```

Note that we have 16 active health checks and 22 waiters. We can see that no health checks are waiting more than 2.5s and all are running in about 1-2.5 seconds. This was exercised with the following command:

```console
$ for i in $(seq 1 50); do docker run -d --health-cmd 'sh -c "sleep 1"' --health-interval 1s --health-timeout 10s ubuntu sleep 1000000; done
```

File: daemon/health.go [hunk 4]
```diff
@@ -188,15 +188,26 @@ func handleProbeResult(d *Daemon, c *container.Container, result *types.Healthch
...
 			return
 		case <-time.After(probeInterval):
-			logrus.Debugf("Running health check for container %s ...", c.ID)
 			startTime := time.Now()
+			healthChecksWaiters.Inc()
 			ctx, cancelProbe := context.WithTimeout(context.Background(), probeTimeout)
+			if err := d.healthcheckLimiter.Acquire(ctx, 1); err != nil {
+				// if we see this, we have ourselves a nasty bug
+				logrus.WithError(err).Warnf("Healthcheck semaphore acquisition failed", c.ID)
+				continue loop
```
Comment (fcrisciani) [inline daemon/health.go]: do you want to decrement here `healthChecksWaiters.Dec()` ?

Comment (fcrisciani) [reply daemon/health.go]: like before continue of course :)

Comment (fcrisciani) [reply daemon/health.go]: maybe a stats heathcheck dropped or denied something like that to capture the ones rejected

**Review by vdemeester — APPROVED** [review]
LGTM :frog:

**Comment by stevvooe** [comment]
I am not sure what is going on with the panic here:

```
22:42:32 ----------------------------------------------------------------------
22:42:32 FAIL: check_test.go:365: DockerSwarmSuite.TearDownTest
22:42:32 
22:42:32 check_test.go:370:
22:42:32     d.Stop(c)
22:42:32 daemon/daemon.go:389:
22:42:32     t.Fatalf("Error while stopping the daemon %s : %v", d.id, err)
22:42:32 ... Error: Error while stopping the daemon d91c1ddc02209 : exit status 130
22:42:32 
22:42:32 
22:42:32 ----------------------------------------------------------------------
22:42:32 PANIC: docker_cli_swarm_test.go:1940: DockerSwarmSuite.TestSwarmClusterEventsService
22:42:32 
22:42:32 [d91c1ddc02209] waiting for daemon to start
22:42:32 [d91c1ddc02209] daemon started
22:42:32 
22:42:32 [d91c1ddc02209] daemon started
22:42:32 Attempt #2: daemon is still running with pid 21705
22:42:32 Attempt #3: daemon is still running with pid 21705
22:42:32 Attempt #4: daemon is still running with pid 21705
22:42:32 [d91c1ddc02209] exiting daemon
22:42:32 ... Panic: Fixture has panicked (see related PANIC)
```

Where is the related panic? Is it possible we are deadlocking on the container or is the failure unrelated?

File: daemon/health.go [hunk 5]
```diff
@@ -188,15 +188,27 @@ func handleProbeResult(d *Daemon, c *container.Container, result *types.Healthch
...
 			logrus.Debugf("Stop healthcheck monitoring for container %s (received while idle)", c.ID)
 			return
 		case <-time.After(probeInterval):
-			logrus.Debugf("Running health check for container %s ...", c.ID)
 			startTime := time.Now()
+			healthChecksWaiters.Inc()
 			ctx, cancelProbe := context.WithTimeout(context.Background(), probeTimeout)
+			if err := d.healthcheckLimiter.Acquire(ctx, 1); err != nil {
+				healthChecksWaiters.Dec()
+				// if we see this, we have ourselves a nasty bug
```
Comment (cpuguy83) [inline daemon/health.go]: Maybe a counter for failures here?

Comment (cpuguy83) [reply daemon/health.go]: Doesn't necessarily need to be added in this PR unless you push another update.

**Review by cpuguy83 — COMMENTED** [review]
A little worried about the size of the limit.
Given that there can be many goroutines just waiting on the probe to finish, it seems like there could be a lot more goroutines waiting for the probe to actually complete rather than being hung up in pipe create/delete.

Another thought, do you think it would be better to move this limiter to libcontainerd?

**Comment by thaJeztah** [comment]
> Where is the related panic? Is it possible we are deadlocking on the container or is the failure unrelated?

I downloaded the bundle for that build; here's the daemon log of that panic

[docker.log.txt](https://github.com/moby/moby/files/1899794/docker.log.txt)

**Comment by thaJeztah** [comment]
Failures in the last run are flaky tests https://jenkins.dockerproject.org/job/Docker-PRs/48911/console (tracked through https://github.com/moby/moby/issues/36501); 

```
22:40:14 ----------------------------------------------------------------------
22:40:14 FAIL: docker_api_swarm_service_test.go:32: DockerSwarmSuite.TestAPIServiceUpdatePort
22:40:14 
22:40:14 [d6abb7d5c4286] waiting for daemon to start
22:40:14 [d6abb7d5c4286] daemon started
22:40:14 
22:40:14 docker_api_swarm_service_test.go:38:
22:40:14     waitAndAssert(c, defaultReconciliationTimeout, d.CheckActiveContainerCount, checker.Equals, 1)
22:40:14 docker_utils_test.go:452:
22:40:14     c.Assert(v, checker, args...)
22:40:14 ... obtained int = 0
22:40:14 ... expected int = 1
22:40:14 
22:40:14 [d6abb7d5c4286] exiting daemon
22:40:26 

23:05:16 ----------------------------------------------------------------------
23:05:16 FAIL: docker_cli_swarm_test.go:1610: DockerSwarmSuite.TestSwarmPublishDuplicatePorts
23:05:16 
23:05:16 [d6a0be587333d] waiting for daemon to start
23:05:16 [d6a0be587333d] daemon started
23:05:16 
23:05:16 docker_cli_swarm_test.go:1618:
23:05:16     // make sure task has been deployed.
23:05:16     waitAndAssert(c, defaultReconciliationTimeout, d.CheckActiveContainerCount, checker.Equals, 1)
23:05:16 docker_utils_test.go:452:
23:05:16     c.Assert(v, checker, args...)
23:05:16 ... obtained int = 0
23:05:16 ... expected int = 1
23:05:16 
23:05:16 [d6a0be587333d] exiting daemon
23:05:19 
```

**Comment by stevvooe** [comment]
@thaJeztah Looking at that dump, I don't see anything blocking on the health check code, so I think these failures are unrelated.

I am open to secondary analysis, but I think we should move forward with this PR.

File: daemon/health.go [hunk 6]
```diff
@@ -188,15 +188,27 @@ func handleProbeResult(d *Daemon, c *container.Container, result *types.Healthch
...
 	for {
 		select {
 		case <-stop:
 			logrus.Debugf("Stop healthcheck monitoring for container %s (received while idle)", c.ID)
 			return
 		case <-time.After(probeInterval):
-			logrus.Debugf("Running health check for container %s ...", c.ID)
 			startTime := time.Now()
+			healthChecksWaiters.Inc()
 			ctx, cancelProbe := context.WithTimeout(context.Background(), probeTimeout)
```
Comment (tonistiigi) [inline daemon/health.go]: Blocking on the limiter shouldn't have an effect on probe timeout.

**Review by tonistiigi — COMMENTED** [review]
I'm not sure I understand why we need an arbitrary limit on deamon level instead of just making sure that a healthcheck per container never runs twice or starts before everything related to the previous one has been completely cleaned up. If this is true it should correctly slow down high load and I'm not sure how it can cause any issues.

File: daemon/health.go (hunk 6 above)
Comment (stevvooe) [reply daemon/health.go]: Why not? This cost would be incurred if it were dogpiling syscalls or waiting on a semaphore.

**Comment by stevvooe** [comment]
@tonistiigi The issue is roughly that by limiting the number of concurrent health checks, we limit the total load spend performing them. If healthchecks timeout while waiting to run, they'll skip a run as a failure and just get back in line for another Go at it, limiting the total load available for health checks.

File: daemon/health.go (hunk 6 above)
Comment (tonistiigi) [reply daemon/health.go]: Probe timeout is what `Dockerfile` author or user starting a container sets depending on the specifics of the application and the health check command. Eg. a reliable healthcheck shouldn't have a long timeout. These users can't predict any scheduling constraints or how the configuration of other containers will start to have an effect on this value.

**Comment by stevvooe** [comment]
Per @tonistiigi's concern, I've added a comment describing the unfortunate situation with health checks.

@cpuguy83 Added the requested counter.

File: daemon/health.go (hunk 5 above)
Comment (stevvooe) [reply daemon/health.go]: Added

**Comment by tonistiigi** [comment]
LGTM

**Review by fcrisciani — APPROVED** [review]
LGTM

**Comment by cpuguy83** [comment]
I'm still trying to reproduce this issue without much luck... hitting lock contention on the buffer pool more than any issue with fifos.

I do have concerns with this in that puts a relatively low limit on healthchecks which may take a fair mount of time (even a couple of seconds for a daemon run lots of healthchecks will back things up) and can easily cause starvation (healthchecks for a specific container(s) never able to run) basically  moving the problem elsewhere.

I can definitely see that there are some race conditions involved with the exit event processing triggering a new healthcheck meanwhile the process (along with fifos) hasn't been cleaned up.... however what's concerning here is the open fifos which are deleted. After the better part of a day looking at this I'm not sure how that could happen outside of exec ID's being recycled + running into the race.

**Comment by thaJeztah** [comment]
Janky failure probably flaky https://jenkins.dockerproject.org/job/Docker-PRs/48997/console

```
02:57:50 FAIL: docker_api_swarm_test.go:297: DockerSwarmSuite.TestAPISwarmLeaderElection
02:57:50 
02:57:50 [dfad20e49e65a] waiting for daemon to start
02:57:50 [dfad20e49e65a] daemon started
02:57:50 
02:57:50 [d2ead159a593f] waiting for daemon to start
02:57:50 [d2ead159a593f] daemon started
02:57:50 
02:57:50 [d774b6241da03] waiting for daemon to start
02:57:50 [d774b6241da03] daemon started
02:57:50 
02:57:50 [dfad20e49e65a] exiting daemon
02:57:50 assertion failed: error is not nil: Error response from daemon: rpc error: code = DeadlineExceeded desc = context deadline exceeded
02:57:50 [d2ead159a593f] exiting daemon
02:57:50 [d774b6241da03] exiting daemon
02:58:10 
```

**Comment by jahkeup** [comment]
@stevvooe, it sounds like there are remaining concerns from @cpuguy83. Are there on going investigations to determine impact that are not recorded here? Is the mentioned FIFO deletion issue being tracked in another issue or perhaps in a containerd repo issue? Is there anything I can do to help test or experiment further?

**Comment by cpuguy83** [comment]
As far as I am aware, the FIFO issues are mitigated by the timeout on fifo open that was added in containerd.
I'm not sure this is needed as an extra mitigation.

**Comment by cpuguy83** [comment]
Those changes were shipped with Docker 18.03.1, btw.

**Comment by jahkeup** [comment]
I did see that the containerd mitigation was included with the recent release, however the conversations I had with @stevvooe on the community Slack during the initial diagnosis and containerd patch testing seemed to indicate that a code change in moby (ie: this change) was needed. Please correct me if I misunderstood and if this PR is considered wholly unneeded with the release of 18.03.1.

**Comment by cpuguy83** [comment]
@jahkeup From what I can tell there are a number of issues related to the same cause and all seem to be mitigated by the containerd change included in 18.03.1.

**Comment by cpuguy83** [comment]
Ok, let's close this since this appears to be mitigated by changes in containerd, which are available in 18.03.1.

Thanks!

**Pull request closed without merging**